File: app/src/main/java/com/heinika/designpattern/abstractfactory/AbstractFactoryPatternDemo.java
package com.heinika.designpattern.abstractfactory;

import com.heinika.designpattern.abstractfactory.color.Red;
import com.heinika.designpattern.abstractfactory.shape.Square;

public class AbstractFactoryPatternDemo {
    public void main(){
        FactoryProducer factoryProducer = new FactoryProducer();
        ColorFactory colorFactory = (ColorFactory) factoryProducer.getFactory(ColorFactory.class.getSimpleName());
        ShapeFactory shapeFactory = (ShapeFactory) factoryProducer.getFactory(ShapeFactory.class.getSimpleName());
        Red red = (Red) colorFactory.getColor(Red.class.getSimpleName());
        red.fill();
        Square square = (Square) shapeFactory.getShape(Square.class.getSimpleName());
        square.draw();
    }
}

File: app/src/main/java/com/heinika/designpattern/kotlin/abstractfactory/AbstractFactoryPatternDemo.kt
package com.heinika.designpattern.kotlin.abstractfactory

import com.heinika.designpattern.kotlin.abstractfactory.color.Red
import com.heinika.designpattern.kotlin.abstractfactory.shape.Square


fun main() {
    val factory = FactoryProducer()
    val shapeFactory = factory.getFactory(ShapeFactory::class.java)
    val square = shapeFactory!!.getShape(Square::class.java)
    square!!.draw()
    val colorFactory = factory.getFactory(ColorFactory::class.java)
    val red = colorFactory!!.getColor(Red::class.java)
    red!!.fill()
}

class AbstractFactoryPatternDemo{}